pkg/kv/kvserver/store.go

```diff
@@ -2936,7 +2936,29 @@ func (s *Store) getRootMemoryMonitorForKV() *mon.BytesMonitor {
 func WriteClusterVersion(
 	ctx context.Context, eng storage.Engine, cv clusterversion.ClusterVersion,
 ) error {
-	return storage.MVCCPutProto(ctx, eng, nil, keys.StoreClusterVersionKey(), hlc.Timestamp{}, nil, &cv)
+	err := storage.MVCCPutProto(ctx, eng, nil, keys.StoreClusterVersionKey(), hlc.Timestamp{}, nil, &cv)
+	if err != nil {
+		return err
+	}
+
+	// The storage engine sometimes must make backwards incompatible
+	// changes. However, the store cluster version key is a key stored
+	// within the storage engine, so it's unavailable when the store is
+	// opened.
+	//
+	// The storage engine maintains its own minimum version on disk that
+	// it may consult it before opening the Engine. This version is
+	// stored in a separate file on the filesystem. For now, write to
+	// this file in combination with the store cluster version key.
+	//
+	// This parallel version state is a bit of a wart and an eventual
+	// goal is to replace the store cluster version key with the storage
+	// engine's flat file. This requires that there are no writes to the
```

Comment:
nit: we don't write to the engine today (at least according to my reading of the code). What we would need to change is that the engine is either created only as part of bootstrap/join (this is not done today due to plumbing reasons, some of which historical) or "unwriteable" before bootstrap/join (slightly less clean solution, but gives the right guarantees). We can effectively "enforce" (at least heuristically) the latter invariant today, which I would ask you to consider.

In https://github.com/cockroachdb/cockroach/blob/e915cb0aef63e498bdc5ff86807cfb8f6f0fe21a/pkg/server/init.go#L495 and https://github.com/cockroachdb/cockroach/blob/e915cb0aef63e498bdc5ff86807cfb8f6f0fe21a/pkg/server/init.go#L517 we should check that all engines (they are all in `unitializedEngines`) are completely empty, i.e. have never been issued a write op to (or whichever other condition you like most - the goal is to make sure that the engine is pretty much as good as one that was *just* created). If they are not, an error is returned. If this fails in any test, we'll know that there is a problem.

Reply:
I might be misreading, but I think this call will at least write to the store version key in advance of boostrapping/joining: https://github.com/cockroachdb/cockroach/blob/e915cb0aef63e498bdc5ff86807cfb8f6f0fe21a/pkg/server/server.go#L1233-L1237 

If we eventually begin to use the separate storage version file exclusively, it seems like we could assert the engine is completely empty. I'll do some more sleuthing; maybe we should just assert that the engine contains no keys other than the store cluster key.